iroh-bytes/src/export.rs

```diff
@@ -0,0 +1,129 @@
+//! Functions to export data from a store
+
+use std::path::PathBuf;
+
+use anyhow::Context;
+use bytes::Bytes;
+use iroh_base::rpc::RpcError;
+use serde::{Deserialize, Serialize};
+use tracing::trace;
+
+use crate::{
+    format::collection::Collection,
+    store::{ExportMode, MapEntry, Store as BaoStore},
+    util::progress::{IdGenerator, ProgressSender},
+    Hash,
+};
+
+/// Export a hash to the local file system.
+///
+/// This exports a single hash, or a collection `recursive` is true, from the `db` store to the
+/// local filesystem. Depending on `mode` the data is either copied or reflinked (if possible).
+///
+/// Progress is reported as [`ExportProgress`] through a [`ProgressSender`]. Note that the
+/// [`ExportProgress::AllDone`] event is not emitted from here, but left to an upper layer to send,
+/// if desired.
+pub async fn export<D: BaoStore>(
+    db: &D,
+    hash: Hash,
+    out: PathBuf,
+    recursive: bool,
+    mode: ExportMode,
+    progress: impl ProgressSender<Msg = ExportProgress> + IdGenerator,
+) -> anyhow::Result<()> {
+    let path = PathBuf::from(&out);
+    if recursive {
```

Comment:
should we error if neither `recursive` nor a parent path is found?

Reply:
The non-recursive case always exports only to a single file, and the parent path is only used to create the parent as a directory if it doesn't exist. So I think we can simply only do the mkdir if parent is set, but always export? Exporting a blob to `/myfile` is a valid usecase I think, no reason to disallow it?

Reply:
I pushed a commit that both reduces code duplication and implements the above (allow exports to a file in root).